Firestore/core/src/api/load_bundle_task.cc

```diff
@@ -27,6 +27,10 @@ namespace firebase {
 namespace firestore {
 namespace api {
 
+LoadBundleTask::~LoadBundleTask() {
+  std::lock_guard<std::mutex> lock(mutex_);
```

Comment:
Question: can you describe a scenario where this lock is helpful?

Reply:
In `TEST_F(LoadBundleTaskTest, ProgressObserverCanRemoveObserver)` for example:

for the last call of `queue.push(3)` happen only user thread. The testing thread could start checking `EXPECT_EQ` right after this call and they will all pass.

Now the user thread could move on to `task.RemoveObserver(2)`, holding the lock, but the testing thread will try to deconstruct `task` right after the checks, hence racing to write the underlying fields.

Reply:
Ah, so sounds like it's to make sure the destructor cannot proceed while an observer is being removed, right? And the problem is that the test can finish in-between `queue.push(3)` and `task.RemoveObserver(2)`?

I think we usually solved this in tests via a `std::promise` that is fulfilled by the callback while the test waits for the promise to be fulfilled before it is finished. This approach seems to work as well. One thing to mention is that in C++ it's usually desirable to define all the copy/move operations and the destructor together (rule of 5). To be pedantic, you would probably need to protect the handle and the observers by a mutex while copying and moving as well.

Reply:
As discussed, made it no copy-able.